67_RayQueryGeometry/main.cpp

```diff
@@ -0,0 +1,1010 @@
+
+				blasBuildInfo.dstAS = gpuBlas.get();
+				blasBuildInfo.scratch.buffer = scratchBuffer;
+				blasBuildInfo.scratch.offset = 0u;
+
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo buildRangeInfos[1u];
+				buildRangeInfos[0].primitiveCount = trisCount;
+				buildRangeInfos[0].primitiveByteOffset = 0u;
+				buildRangeInfos[0].firstVertex = 0u;
+				buildRangeInfos[0].transformByteOffset = 0u;
+				IGPUBottomLevelAccelerationStructure::BuildRangeInfo* pRangeInfos[1u];
+				pRangeInfos[0] = &buildRangeInfos[0];
+
+				cmdbuf->buildAccelerationStructures({ &blasBuildInfo, 1 }, pRangeInfos);
+			}
+
+			{
+				SMemoryBarrier memBarrier;
+				memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+				memBarrier.srcAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_WRITE_BIT;
+				memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::ACCELERATION_STRUCTURE_BUILD_BIT;
+				memBarrier.dstAccessMask = ACCESS_FLAGS::ACCELERATION_STRUCTURE_READ_BIT;
+				cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, { .memBarriers = {&memBarrier, 1} });
+			}
+
+			// compact blas
+			IQueryPool::SCreationParams qParams{ .queryCount = 1, .queryType = IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE };
+			smart_refctd_ptr<IQueryPool> queryPool = m_device->createQueryPool(std::move(qParams));
+
+			uint32_t queryCount = 0;
+			const IGPUAccelerationStructure* ases[1u] = { gpuBlas.get() };
+			cmdbuf->writeAccelerationStructureProperties({ ases, 1}, IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE, queryPool.get(), queryCount++);
+
+			size_t asSizes[1];
+			m_device->getQueryPoolResults(queryPool.get(), 0, queryCount, asSizes, sizeof(size_t), IQueryPool::WAIT_BIT);
+			
+			auto cleanupBlas = gpuBlas;
+			{
+				IGPUBuffer::SCreationParams params;
+				params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
+				params.size = asSizes[0];
+				smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
+
+				IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
+				blasParams.bufferRange.buffer = asBuffer;
+				blasParams.bufferRange.offset = 0u;
+				blasParams.bufferRange.size = asSizes[0];
+				blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
+				gpuBlas = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
+			}
+
+			IGPUBottomLevelAccelerationStructure::CopyInfo copyInfo;
+			copyInfo.src = cleanupBlas.get();
+			copyInfo.dst = gpuBlas.get();
+			copyInfo.mode = IGPUBottomLevelAccelerationStructure::COPY_MODE::COMPACT;
+			cmdbuf->copyAccelerationStructure(copyInfo);
+
+			// build top level AS
+			{
+				const uint32_t instancesCount = 1;	// TODO: temporary for now
+				IGPUTopLevelAccelerationStructure::DeviceInstance instances[instancesCount];
+				core::matrix3x4SIMD identity;
+				instances[0].blas.deviceAddress = gpuBlas->getCreationParams().bufferRange.buffer->getDeviceAddress();
```

Comment:
the TLAS references a BLAS and cannot happen at the same time as the BLAS build (or copies) of the BLASes its referencing, hence you need another pipeline barrier here.